### PR TITLE
Inject git commit sha during build process.

### PR DESCRIPTION
**Pull request by hpidcock**
## Description of change

Inject git commit sha during build process.

- juju version --all now prints the git commit, git tree state, version
and compiler
- show-controller now prints the git commit sha (and reports correct
version)
- juju/juju/version package holds the git commit and tree state at build
time.

### juju version --all
```
$ juju version --all
version: 2.6.6-disco-amd64
git-commit: fe41ff899d9c80ec34d40c14ac9b9a9300071095
git-tree-state: clean
compiler: gc
```

### juju show-controller
```
$ juju show-controller
microk8s-localhost:
  details:
    uuid: 7f0dba8e-2170-4554-8cb0-13a916648ff2
    controller-uuid: 7f0dba8e-2170-4554-8cb0-13a916648ff2
    api-endpoints: ['10.152.183.108:17070']
    cloud: microk8s
    region: localhost
    agent-version: 2.6.6
    agent-git-commit: fe41ff899d9c80ec34d40c14ac9b9a9300071095
    controller-model-version: 2.6.6
    mongo-version: 4.0.6
...
```

## Documentation changes

Current workflow should remain unchanged.

## Bug reference

https://bugs.launchpad.net/juju/+bug/1832784

### CODE REVIEW

**Comment by hpidcock** [comment]
!!build!!

**Comment by anastasiamac** [comment]
[inject sha into juju binaries](https://trello.com/c/A7Mji1r7/808-inject-sha-into-juju-binaries)

File: apiserver/facades/client/controller/controller.go [hunk 1]
```diff
@@ -185,6 +201,22 @@ func (c *ControllerAPI) checkHasAdmin() error {
 	return nil
 }
 
+// ControllerVersion isn't on the v7 API.
+func (c *ControllerAPIv7) ControllerVersion() {}
```
Comment (anastasiamac) [inline apiserver/facades/client/controller/controller.go]: Current project preference:
```
func (c *ControllerAPIv7) ControllerVersion(_, _ struct{}) {}
```

File: apiserver/params/controller.go [hunk 2]
```diff
@@ -106,3 +106,10 @@ const (
 	GrantControllerAccess  ControllerAction = "grant"
 	RevokeControllerAccess ControllerAction = "revoke"
 )
+
+// ControllerVersionResults holds the results from an api call
+// to get the controller's version information.
+type ControllerVersionResults struct {
+	Version   string `json:"version"`
```
Comment (anastasiamac) [inline apiserver/params/controller.go]: Please add comments to individual properties too.

**Comment by hpidcock** [comment]
Reduced the scope of this PR and moved the rest to https://github.com/juju/juju/pull/10435

**Comment by hpidcock** [comment]
!!BUILD!!

**Comment by hpidcock** [comment]
!!build!!

**Comment by hpidcock** [comment]
!!build!!

**Comment by hpidcock** [comment]
$$merge$$

**Pull request merged**